#! /bin/sh /usr/share/dpatch/dpatch-run
## endless_loop.dpatch by Ronald Bultje <rbultje@ronald.bitfreak.net>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: If the ID3v2 tag contains string list fields with an UTF16 string whose 
## DP: length is an odd number of bytes (which is effectively a broken string, 
## DP: but should still be parsed correctly), libid3tag ends up in an endless 
## DP: loop allocating memory until it can't allocate more memory or the 
## DP: process is killed.

@DPATCH@
--- /home/debian/mad/libid3tag-0.15.1b/libid3tag-0.15.1b/utf16.c	2005-01-07 12:35:28.622883237 +0100
+++ libid3tag-0.15.1b/utf16.c	2005-01-07 12:35:36.603673548 +0100
@@ -250,6 +250,8 @@
   id3_ucs4_t *ucs4;
 
   end = *ptr + (length & ~1);
+  if (end == *ptr)
+    return 0;
 
   utf16 = malloc((length / 2 + 1) * sizeof(*utf16));
   if (utf16 == 0)
